[JSC] `Math.sumPrecise` Enforces Maximum Iterations and Throws `RangeError`

Gumichocopengin8 opened the pull request:
#### be9083436a734d60cc4d14d97c9bfea87cd57b50
<pre>
[JSC] `Math.sumPrecise` Enforces Maximum Iterations and Throws `RangeError`
<a href="https://bugs.webkit.org/show_bug.cgi?id=298174">https://bugs.webkit.org/show_bug.cgi?id=298174</a>

Reviewed by Sosuke Suzuki.

Previously, `Math.sumPrecise` did not throw a `RangeError`
when the iteration count exceeded `2 ^ 53 - 1 (9007199254740991)`.
This patch ensures that the maximum iteration limit is enforced
according to TC39 spec [1].

[1]: <a href="https://tc39.es/ecma262/#sec-math.sumprecise">https://tc39.es/ecma262/#sec-math.sumprecise</a>

* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/299384@main">https://commits.webkit.org/299384@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92612066bdc4ac6c5964e1459147a83991dccf6a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118849 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125038 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70911 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47112 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90198 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59713 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121802 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31262 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106557 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70705 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24671 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68697 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110967 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100705 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24858 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128089 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117363 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45756 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34555 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98854 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46122 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102777 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98634 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25069 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44081 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22088 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42316 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45626 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51304 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146061 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45091 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37569 "Found 1 new JSC binary failure: testapi, Found 18646 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48436 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46776 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -459,13 +459,19 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncSumPrecise, (JSGlobalObject* globalObject,
+        uint64_t count = 0ULL;
+        forEachInIterable(globalObject, iterable, [&sum, &count](VM& vm, JSGlobalObject* globalObject, JSValue value) {
             auto scope = DECLARE_THROW_SCOPE(vm);
+            if (count >= maxSafeIntegerAsUInt64()) {
```
Add `[[unlikely]]`

sosukesuzuki commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -459,13 +459,19 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncSumPrecise, (JSGlobalObject* globalObject,
+                throwRangeError(globalObject, scope, "Math.sumPrecise exceeded maximum iterations"_s);
+                return;
+            }
             if (!value.isNumber()) {
```
Add `[[unlikely]]`. (I know this line is not related to this PR, but I think we can do it here)

sosukesuzuki commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -459,13 +459,19 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncSumPrecise, (JSGlobalObject* globalObject,
 
     auto calculatePreciseSum = [&](auto& sum) {
-        forEachInIterable(globalObject, iterable, [&sum](VM& vm, JSGlobalObject* globalObject, JSValue value) {
+        uint64_t count = 0ULL;
```
I think `ULL` is unnecessary here

sosukesuzuki reviewed (commented):
If it's not covered by test262, please add tests.

sosukesuzuki reviewed (approved):
r=me

Gumichocopengin8 committed:
[JSC] `Math.sumPrecise` Enforces Maximum Iterations and Throws `RangeError`
https://bugs.webkit.org/show_bug.cgi?id=298174

Reviewed by Sosuke Suzuki.

Previously, `Math.sumPrecise` did not throw a `RangeError`
when the iteration count exceeded `2 ^ 53 - 1 (9007199254740991)`.
This patch ensures that the maximum iteration limit is enforced
according to TC39 spec [1].

[1]: https://tc39.es/ecma262/#sec-math.sumprecise

* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/299384@main
